### ISSUE TITLE
Add-Type : Could not load file or assembly netstandard

### ISSUE THREAD

**Issue by abelal83** [PowerShell/PowerShell#10955]
Not sure what I'm doing wrong here...

I downloaded the application.insightst nuget package v 2.11.0, unblocked all files and extraced to a folder. I have since been trying to use `Add-Type -Path Microsoft.ApplicationInsights.dll` but it simply doens't work. Tried the version in netstandard1.3 and netstandard2.0.

Is it the case netstandard compiled dll's are not supported in powershell core 6.2? I'm ablke to use Add-Type suceesfully with dll's compiled against netcoreapp.

**Comment by SteveL-MSFT** [comment]
PowerShell Core already uses AppInsights.  If you look in $PSHOME, you'll see the assembly there.  Since pwsh loads it already, you wouldn't be able to load a different version anyways.  But you can use the one that comes with pwsh.

**Comment by abelal83** [comment]
@SteveL-MSFT thank you very much for your prompt response!
I had no idea it was there, this should get me working with that.

So am I wrong in saying netstandard compiled dlls can't be used with add-type?

**Comment by SteveL-MSFT** [comment]
netstandard compiled dlls are just like any other assembly from PowerShell perspective and can be used, but there are some limitations like having an existing assembly loaded you can't currently have two versions of the same assembly loaded at the same time (first one wins).

**Comment by johnwc** [comment]
@SteveL-MSFT So this is actually an issue I've run into before, where two modules want to use the same library but different versions. In this AppInsight scenario, how do we work around this? Are we forced to downgrade our c# psmodule to using the same version as pwsh?

**Comment by rjmholt** [comment]
PowerShell loads everything into the same Assembly Load Context / AppDomain, so to get both modules working at once they must have the same dependency versions.

If you're consuming modules with the same dependency in different versions, your choices are:

- Downgrade module versions to align dependency versions
- Load the modules into different PSSessions, so that at least one of them is loaded remotely. You will get deserialised PSObjects back from commands, but it should work.

I'm afraid that as a module consumer, there's simply no way to load two modules with conflicting assemblies and have both give back full, live .NET objects. That's a limitation in .NET.

If you are the module author you have some choices:

- Align to a common dependency version (for example, the Newtonsoft.Json version PowerShell uses)
- Implement a fresh Assembly Load Context within your own module to isolate your dependencies.

We've experimented with this second solution in the case of the PowerShellEditorServices module, where PSES loads its dependencies into a new ALC in PS 6.2/7, but it can cause subtle issues. We may eventually be able to deliver a general solution for modules in PowerShell itself, but it's a large and complex work item and we have no timeframe for even a design proposal.

The other problem is that this only works in PS 6.2+. In Windows PowerShell, which uses .NET Framework, it may be possible for module authors to isolate their dependencies in another AppDomain, but that's a serialisation boundary, meaning the solution becomes more complex again.

**Comment by johnwc** [comment]
@rjmholt Our main issue, is that we are needing to reference the AppInsight library, and are not able to reference the latest release. As pwsh itself references is as well and refences an older version.

**Comment by rjmholt** [comment]
> As pwsh itself references is as well and refences an older version.

Yeah, this issue also occurs for Newtonsoft.Json. You must use the lowest version of AppInsights that any of your supported PowerShells targets, or otherwise use a mechanism like Assembly Load Contexts to isolate your dependency from PowerShell. Whether the dependency is shared with another module or PowerShell itself currently doesn't matter; the symptoms and solution remain the same.

**Comment by johnwc** [comment]
Do you have an example of using this approach?

**Comment by rjmholt** [comment]
> Do you have an example of using this approach?

If you mean the version approach, [it's what PSScriptAnalyzer does for Newtonsoft.Json](https://github.com/PowerShell/PSScriptAnalyzer/blob/26f34b56466cf30862d2ffadf0a5f3c9347b3936/Rules/Rules.csproj#L14). This is, by far, the easiest way to address the issue.

Using an assembly load isolation mechanism is really something I would recommend only if you have no other choice (like if indirect dependencies force you to resolve a version conflict).

The only example I know of is in PowerShell Editor Services, [where we used an Assembly Load Context to load our dependencies in .NET Core](https://github.com/PowerShell/PowerShellEditorServices/blob/fc1a95a689aa1fdd2fc9b94f84004e5b2c3ec6a5/src/PowerShellEditorServices.Hosting/EditorServicesLoader.cs#L80-L152) so that other modules with conflicting dependencies will still work. However, because of the constraints of AppDomains (namely that they're a remoting boundary), our solution in .NET Core is much better than the .NET Framework one (which is to use Assembly.LoadFile()).

There's some coverage of the topic in this thread: https://github.com/PowerShell/PowerShell/issues/11571#issuecomment-574384979

In such a case, you'll need to:

- Add a new assembly to your codebase that is targeted to both .NET Framework and .NET Core (i.e. it will need two TargetRuntimes)
- This assembly will need to provide a common API for your isolated dependencies for the rest of your library/module, but must do so without exporting any of the types of those dependencies; it needs to totally wrap those dependencies and only use your own or PowerShell-common types (i.e. types you've defined in your library, or types that are available in a clean PowerShell session -- ones that come from assemblies in $PSHome).
- In the .NET Framework part, you'll need to create a new AppDomain, load the required dependencies into it, implement a remote proxy or some other way of marshalling information between your dependency AppDomain and the main AppDomain into which the module is loaded. You could also try Assembly.LoadFile, but there may be caveats with that. I haven't tried the AppDomain approach myself because the serialisation boundary made it a bad choice in our case.
- .NET Core is much easier; you implement an AssemblyLoadContext, write it so that it loads your dependencies from a particular location (like a separate directory), and add a resolve event to redirect those types back to your facade assembly
- Then the facade assembly just takes inputs and routes them through to the actual dependencies over the isolating abstraction layer

The bottom line is that a workaround is *possible*, but it's a lot of work and you should really evaluate how much you need to not align to a common version.
